functions.py

```diff
@@ -0,0 +1,757 @@
+            circuit.add(error_gate(C[i], err))
+        if int(c[3][i]) == 1:
+            circuit.add(gates.X(D[i]))
+            circuit.add(error_gate(D[i], err))
+    circuit.add(diffuser(A+B, x, err))
+    return circuit
+
+
+def check_hash(q, message, h, constant_1, constant_2, rot):
+    """Check if a given output message is a preimage of a given hash value.
+    Args:
+        q (int): number of qubits of a site in the permutation matrix.
+        message (str): output message that we want to check.
+        h (str): hash value that one wants to find preimages of.
+        constant_1 (int): constant that defines the hash construction.
+        constant_2 (int): constant that defines the hash construction.
+        rot (list): characterization of the rotation part of the algorithm.
+            
+    Returns:
+        True of False if the message correspongs to a preimage.
+    """
+    n = int(len(h))
+    m1 = 0
+    m2 = 0
+    for i in range(q):
+        m1 += int(message[i])*2**(q-1-i)
+        m2 += int(message[q+i])*2**(q-1-i)
+    a = "{0:0{bits}b}".format(0, bits=q)
+    b = "{0:0{bits}b}".format(0, bits=q)
+    c = "{0:0{bits}b}".format(constant_1, bits=q)
+    d = "{0:0{bits}b}".format(constant_2, bits=q)
+    for i in range(10):
+        a, c = chacha_qr(q, a, c, rot)
+        b, d = chacha_qr(q, b, d, rot)
+        a, d = chacha_qr(q, a, d, rot)
+        b, c = chacha_qr(q, b, c, rot)
+    A = 0
+    B = 0
+    for i in range(q):
+        A += int(a[i])*2**(q-1-i)
+        B += int(b[i])*2**(q-1-i)
+    A = A ^ m1
+    B = B ^ m2
+    a = "{0:0{bits}b}".format(A, bits=q)
+    b = "{0:0{bits}b}".format(B, bits=q)
+    for i in range(10):
+        a, c = chacha_qr(q, a, c, rot)
+        b, d = chacha_qr(q, b, d, rot)
+        a, d = chacha_qr(q, a, d, rot)
+        b, c = chacha_qr(q, b, c, rot)
+    output = (a+b)[:n]
+    return h == output
+
+
+def grover(q, constant_1, constant_2, rot, h, grover_it, nshots=100, err=0):
+    """Run the full Grover's search algorithm to find a preimage of a hash function.
+    Args:
+        q (int): number of qubits of a site in the permutation matrix.
+        constant_1 (int): constant that defines the hash construction.
+        constant_2 (int): constant that defines the hash construction.
+        rot (list): characterization of the rotation part of the algorithm.
+        h (str): hash value that one wants to find preimages of.
+        grover_it (int): number of Grover steps to be performed.
+        nshots (int): number of shots to perform on the final state.
+        err (float): error probability.
+            
+    Returns:
+        result (dict): counts of the output generated by the algorithm
+    """
+    A, B, C, D, x, ancilla, circuit, qubits = create_qc(q)
+    c1, c2, c3, c4 = initial_step(q, constant_1, constant_2, rot)
+    c = []
+    c.append(c1)
+    c.append(c2)
+    c.append(c3)
+    c.append(c4)
+    circuit.add(start_grover(A+B, ancilla))
```

Comment:
I think the error is missing:
```suggestion
    circuit.add(start_grover(A+B, ancilla, err))
```